ghost/core/test/utils/e2e-framework.js

```diff
@@ -335,6 +341,11 @@ const getAgentsWithFrontend = async () => {
         server: true
     };
     try {
+        // Possible that we still have a running Ghost server from a previous old E2E test
+        // Those tests never stopped the server in the tests manually
+        await stopGhost();
```

Comment:
AFAIR, adding `stopGhost` here would introduce a significant overhead to test suite running time. We deliberately avoided using stopGhost in the e2e framework to speed the tests up. That's why `startGhost` is not an actual HTTP server, it's an express app which we [hook into with the framework](https://github.com/TryGhost/framework/blob/36215e772d3983781771f3507972a18974334894/packages/express-test/lib/request.js#L22) directly to bypass HTTP overhead. 

In which specific cases/orders are the test breaking? Is there an alternative way without forcing the `stopGhost` here?

Reply:
Do note that we don't stop ghost for every E2E test. This is only for E2E tests that still need the legacy Ghost HTTP instance (not the Express app) to access the frontend. And if that happens, we only stop the Ghost **HTTP** server that was created in the tests that use the old E2E framework. Also, I don't think this causes any overhead because inside `stopGhost`,  it checks for `ghostServer` to be running. So it only stops the existing (legacy!) http server if it is needed.

So to give a better overview.
- We have tests that use the legacy e2e-utils file to startGhost. These never stop Ghost, because startGhost in there stops it for them. This means that the Ghost HTTP server is left open when the last legacy test E2E test runs.
- In our new E2E test framework, we normally never start a HTTP server. Unless if we need the frontend. For some reason we can't really access the frontend using the new test agents. But that can be fixed in the future probably. As a temporary workaround we still use `supertest.agent` for that. And that requires a HTTP server. So for a minority of new tests, we still need a HTTP server (for now). In that case we need to stop existing HTTP servers from the legacy E2E tests if it is running.

Since the test ordering is very unpredictable, and the stopGhost server only stops the legacy HTTP server if it is running, I don't see a big problem in keeping this here for now until we are able to switch the new E2E tests that need the frontend server to a TestAgent that doesn't require HTTP.

Reply:
Okay, gotcha! The dots are connected now 😄 

This explanation reminded me of a [traumatic commit](https://github.com/TryGhost/Ghost/commit/26b0bbc623885f65ce770cec2a0daf5291c115d7) I've made last year. As a result of that commit we have [this issue](https://github.com/TryGhost/Toolbox/issues/471), which I think is worth referencing here. That issue mentions:
>  fix the underlying problem of state management when frontend has to be initialized for e2e tests

^ this is pretty much the problem we are solving here, right? 

A more explicit solutions would be making sure we call "ghostStop" in the "after" clause wherever we start ghost with frontend, like [done here](https://github.com/TryGhost/Ghost/commit/26b0bbc623885f65ce770cec2a0daf5291c115d7#diff-ef9a744fed6a74ceed8b2346945c7dfa8cddd752ead8f1522bb1687307b8f0e3R38). It's away to much to track and remember of other engineers to add in their tests when working on new suites. Your solutions seems more universal, given there's no noticeable overhead just calling the 'stopServer' every time. 

I think the reasoning is good behind these changes! Can't think of a better approach to fix those "dirty" frontend starts 😅